pkg/apis/storage/types.go

```diff
@@ -424,3 +445,77 @@ type CSINodeList struct {
 	// items is the list of CSINode
 	Items []CSINode
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIStorageCapacity stores the result of one CSI GetCapacity call.
+// For a given StorageClass, this describes the available capacity in a
+// particular topology segment.  This can be used when considering where to
+// instantiate new PersistentVolumes.
+//
+// For example this can express things like:
+// - StorageClass "standard" has "1234 GiB" available in "topology.kubernetes.io/zone=us-east1"
+// - StorageClass "localssd" has "10 GiB" available in "kubernetes.io/hostname=knode-abc123"
+//
+// The following three cases all imply that no capacity is available for
+// a certain combination:
+// - no object exists with suitable topology and storage class name
+// - such an object exists, but the capacity is unset
+// - such an object exists, but the capacity is zero
+//
+// The producer of these objects can decide which approach is more suitable.
+//
+// This is an alpha feature and only available when the CSIStorageCapacity feature is enabled.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning. It must be
+	// be a DNS subdomain (dots allowed, 253 characters). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>, a generated name, or a reverse-domain name which ends
+	// with the unique CSI driver name.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// NodeTopology defines which nodes have access to the storage
+	// for which capacity was reported. If not set, the storage is
+	// not accessible from any node in the cluster. If empty, the
+	// storage is accessible from all nodes.  This field is
+	// immutable.
+	//
+	// +optional
+	NodeTopology *metav1.LabelSelector
+
+	// The name of the StorageClass that the reported capacity applies to.
+	// This field is immutable.
+	StorageClassName string
```

Comment:
ISTR `StorageClass: ""` is explicitly defined for PV and PVC.  Does it matter here?  Can this be empty?

Reply:
No, not here. The empty string references the storage class with the name "default", which is also what will be used here (the string "default", not the empty string).

Reply:
In PVC, empty string means "no storage class". Nil means default. I think in this particular case, we should always require a StorageClass and not support a default. Also, there is no defaulting of StorageClass names in Kubernetes.

I think in this case, we should make StorageClassName required to be non-empty.

Reply:
> I think in this case, we should make StorageClassName required to be non-empty.

This is how it is implemented already. Let me add a comment which explicitly documents that.

Reply:
Done, update pushed.